src/openms/source/FORMAT/HANDLERS/MzMLHandler.cpp

```diff
@@ -3907,88 +3919,252 @@ namespace OpenMS::Internal
-        {
-          logger_.setProgress(progress++);
-          const ChromatogramType& chromatogram = exp.getChromatograms()[c_idx];
-          writeChromatogram_(os, chromatogram, c_idx, validator);
-          ++stored_chromatograms;
+            // Calculate offset if indexing is enabled 
+            Int64 offset = -1;
+            if (options_.getWriteIndex())
+            {
+                if (!compress) {
+                    offset = static_cast<Int64>(os.tellp());
+                }
+                else { 
+                    offset = counter_filter.characters();
+                }
+            }
+            std::string native_id = exp[s_idx].getNativeID();
+            if (renew_native_ids)
+            {
+                native_id = "scan=" + String(s_idx);
+            }
+            if (options_.getWriteIndex() && offset != -1)
+            {
+                spectra_offsets_.emplace_back(native_id, offset);
+            }
+            writeSpectrum_(*output_stream, exp[s_idx], s_idx, validator, renew_native_ids, dps);
+            stored_spectra++;
         }
-        os << "\t\t</chromatogramList>" << "\n";
-      }
+        *output_stream << "\t\t</spectrumList>\n";
+    }
+  
+    // Write chromatograms
+    if (!exp.getChromatograms().empty())
+    {
+        *output_stream << "\t\t<chromatogramList count=\"" << exp.getChromatograms().size() << "\" defaultDataProcessingRef=\"dp_sp_0\">\n";
+        for (Size c_idx = 0; c_idx < exp.getChromatograms().size(); ++c_idx)
+        {
+            logger_.setProgress(progress++);
 
-      MzMLHandlerHelper::writeFooter_(os, options_, spectra_offsets_, chromatograms_offsets_);
+            // Calculate offset if indexing is enabled
+            Int64 offset = -1;
+            if (options_.getWriteIndex())
+            {
+                if (!compress)
+                {
+                    offset = static_cast<Int64>(os.tellp());
+                }
+                else
+                {
+                    offset = counter_filter.characters();
+                }
+            }
 
-      OPENMS_LOG_INFO << stored_spectra << " spectra and " << stored_chromatograms << " chromatograms stored.\n";
+            if (options_.getWriteIndex() && offset != -1)
+            {
+                chromatograms_offsets_.emplace_back(exp.getChromatograms()[c_idx].getNativeID(), offset);
+            }
+            writeChromatogram_(*output_stream, exp.getChromatograms()[c_idx], c_idx, validator);
+            stored_chromatograms++;
+        }
+        *output_stream << "\t\t</chromatogramList>\n";
+    }
 
-      logger_.endProgress(os.tellp());
+    // Write footer
+    if (options_.getWriteIndex())
+    {
+        MzMLHandlerHelper::writeFooter_(*output_stream, options_, spectra_offsets_, chromatograms_offsets_);
+    }
+    else
+    {
+        std::vector<std::pair<std::string, Int64>> empty;
+        MzMLHandlerHelper::writeFooter_(*output_stream, options_, empty, empty);
     }
```

Comment:
why make a distinction here? It worked before without if/else. `chromatograms_offsets_` should be empty if not index is written?